telethon/network/connection/tcpabridged.py

```diff
@@ -9,34 +9,36 @@ class ConnectionTcpAbridged(Connection):
     only require 1 byte if the packet length is less than
     508 bytes (127 << 2, which is very common).
     """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._codec = AbridgedPacket()
+
     def _init_conn(self):
-        self._writer.write(b'\xef')
+        self._writer.write(self._codec.tag)
```

Comment:
Given that all `Connection`'s can now optionally have a `_codec`, perhaps this should move out to the parent class to avoid repeating ourselves in subclasses. As well as providing a default codec with no tag and not modifying the sent/received data.

Reply:
I also had this idea. What I propose is following:

Connection base class will have `codec` and `obfuscation` static attributes. Subclasses can redefine them by assigning specific `codec` and `obfuscation` classes to these attributes.
`Connection` class will just instantiate them in `__init__` and `connect` and will provide generic `_send` and `_recv` methods that will use `self._codec` and `self._obfuscation` to encode and decode packets.

```
Connection:
    codec = None
    obfuscation = NoopObfuscation
    def __init__(...):
        self._codec = self.codec()
    def connect(..):
        self._obfuscation = self.obfuscation(...)
        # somehow send "header", yet to decide how to choose if use tag from obfuscation or codec
    def _recv():
        return await self._codec.read_packet(self._obfuscation)
    def _send(self, data):
        self._obfuscation.write(self._codec.encode_packet(data))
```

Concrete classes will just redefine this 2 static attributes

```
class ConnectionTcpAbridged(Connection):
    codec = AbridgedCodec

class ConnectionTcpFull(Connection):
    codec = FullCodec

class ConnectionTcpObfuscated(Connection):
    codec = AbridgedCodec
    obfuscation = ObfuscatedIO

class ConnectionTcpMTProtoAbridged(Connection):
    codec = AbridgedCodec
    obfuscation = MTProtoIO

...
```

Or we can ommit obfuscation part in Connection class and have a common subclass that adds obfuscation layer

Reply:
I like this approach! Please do it if you have the time and interest.

Reply:
@Lonami yep, thanks! Will do!